### ISSUE TITLE
Which standard is used for the type definitions?

### ISSUE THREAD

**Issue by danse** [typed-wire/typed-wire#5]
I skimmed through the docs and the code but couldn't easily find out what are you using to encode the type definitions. I want to suggest using [JSON schema](http://json-schema.org/), this way it would be easy to reuse the type definitions with other languages, because everybody could build a generator also outside `typed-wire'. The format is not great, but it is the most commonly used for JSON, as far as i know

**Comment by agrafix** [comment]
Does JSON schema support generics?

**Comment by danse** [comment]
Sorry, i don't understand what you mean. A JSON Schema describes a JSON document. What would be a generic value in a JSON document? Maybe just an unspecified object?

**Comment by paulyoung** [comment]
I don't believe it does, but there are some [validation options](http://json-schema.org/latest/json-schema-validation.html) which may be of some use regarding that.

Perhaps it would be worth looking into [jsonschema-gen](https://github.com/yuga/jsonschema-gen) and [aeson-schema](https://github.com/Fuuzetsu/aeson-schema).

Speaking of standards, have you considered [JSON API](http://jsonapi.org/)?

**Comment by agrafix** [comment]
@danse I would like to be able to express types such as

```
type Chicken {
   legs: Int;
}
type Firefly {
   wings: Int;
}
type Pet<t> {
   animal : t;
   name : Int;
}
```

Which can be used as `Pet<Chicken>` and `Pet<Firefly>`. Now to cover this with JSON schema one would have to write something like

```
{
    "title": "Pet",
    "type": "object",
    "properties": {
        "animal": {
            "type": "object",
            "oneOf": ["Chicken", "Firefly"] // something like this...
        },
        "name": {
            "type": "string"
        }
    },
    "required": ["animal", "name"]
}
```

This works well here, but consider the types

```
type WithId<t> {
   id: Int;
   value: t;
}

enum LinkedList<t> {
   Nil;
   Cons(LinkedListValue<t>);
}

type LinkedListValue<t> {
   value: t;
   next: LinkedList<t>;
}
```

How can I express that any type defined within typed-wire can instantiate `WithId<t>` or `LinkedList<t>`? Should we list all possibilities in there? I'm also not sure how to generate Haskell Code from that JSON-Schema. I think if you really need JSON schema, generating it from the typed-wire definition would be a better option?

@paulyoung JSON API sounds like a good idea, but i think you theoretically could already model it using the existing typed-wire spec?

```
type JsonApi<t> {
   data: ResourceObject<t>;
   meta: MetaObject<t>;
   errors: ErrorObject;
   jsonapi: Option<ApiSpec>;
   links: Option<Links>;
   included: Option<Included>;
}

// ...
```

**Comment by danse** [comment]
I can understand that JSON-Schema is not powerful enough for your needs. It might have other limitations besides the one you already mentioned, or you might not like it for more superficial reasons. Personally i find it cumbersome, still as far as i know it is the most used standard to define a JSON schema.

If you want to go for your own definition language, please accept these suggestions:
- Let the schema language live in its own repo, as it is something possibly bigger than `typed-wire`. People might want to use it with a plethora of other languages or tools... for example a _typewire-language-to-json-schema-converter_ tool
- Keep in mind that JSON Schema exists and people build upon it, and reusing at least some of its parts would be nice. For example you could design your schema language so that it could refer to external JSON schemas, reusing the hyperlinking part. I am not an expert, so this is just an idea

I just wanted to mention this while you are still in an early phase, in order to help with the design. Feel free to close the issue if you want

**Comment by bwestergard** [comment]
@agrafix I quite like the examples of the type declaration format, but would love to see a formal specification. If you write one, I'd be happy to write a Flow-typed javascript extractor generator for it.
